test/gtest/uct/ib/test_devx.cc

```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+
+    /* Drop XGVMI capability */
+    md()->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+
+    /* UMR objects remains uninitialized, exported key is not UMR one */
+    check_umr_init(false, false);
+    check_memh_export(memh, false);
+    destroy_memh(memh);
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+    /* XGVMI capability might be dropped by this point! */
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+    /* Whether mkey is UMR depends on XGVMI capability */
+    check_memh_export(memh, check_xgvmi());
+
+    destroy_memh(memh);
+
+    /* After release of memh, associated UMR mkey is moved back to mkey pool */
+    ASSERT_EQ(check_xgvmi() ? 1 : 0, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi_reuse,
+                     !check_xgvmi_effective())
+{
+    const size_t MEMH_COUNT = 10;
+    uct_ib_mlx5_devx_mem_t *memh[MEMH_COUNT];
+
+    for (size_t i = 0; i < MEMH_COUNT; ++i) {
+        memh[i] = create_memh(1024);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
+        check_memh_export(memh[i], true);
+    }
+
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+
+    for (size_t i = 0; i < MEMH_COUNT; ++i) {
+        destroy_memh(memh[i]);
+    }
+
+    /* After release all allocated UMR keys go back to the mkey pool */
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    /* Now reuse the keys for different handles of different size */
+    const size_t REUSE_COUNT = 5;
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        memh[i] = create_memh(2048);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
+        check_memh_export(memh[i], true);
+    }
+
+    ASSERT_EQ(MEMH_COUNT - REUSE_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        destroy_memh(memh[i]);
+    }
+
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, import_mkey_no_umr,
+                     !check_xgvmi_effective())
+{
+    /* Export any mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
+
+    /* Construct packed mkey without UMR tag */
+    uct_ib_md_packed_mkey_t packed_mkey = {
+        .lkey = exported_memh->exported_lkey & ~UCT_IB_MLX5_MKEY_TAG_UMR,
+        .vhca_id = md()->super.vhca_id
```

Comment:
can we disable UMR export in another way? maybe by configuration?
it's too low-level to use mkey tags in the test

Reply:
These are the options that I see:
1. Add configuration e.g. `UCX_UMR_ENABLE` like you said. IMO this is the least desirable option, because there is no use case in production workflow for this option (right?), so we add it in production code purely for unit testing purposes, which seems wrong.
2. Using mocking library. We would be able to mock internal function call like:
`mock().expectOneCall("uct_ib_mlx5_is_xgvmi_supported").andReturnValue(0);`
3. Create MR of size > 2GB, in this case it will be handled by KSM workflow. Mikhail advised me not to do that in unit test, because it can be an expensive call.
4. Remove UMR tag from the mkey - like is done today. IMO it's ok to do low-level things in unit tests - they are low-level tests by definition.

Do you have a preference?

Reply:
> 1. Add configuration e.g. `UCX_UMR_ENABLE` like you said. IMO this is the least desirable option, because there is no use case in production workflow for this option (right?), so we add it in production code purely for unit testing purposes, which seems wrong.

Actually it can be useful to add option to enable/disable this feature in case we find SW/HE issues with this flow. Since it can be disabled without hurting functionality.

Reply:
ok, adding an option then